File: backend/src/services/subsidiaryLocaleService.ts
import { AppDataSource } from "../config/data-source";
import { SubsidiaryLocale } from "../entities/SubsidiaryLocale";
import { ConflictError } from "../utils/errors";

/** Every master locale row, every subsidiary — the admin management view. */
export function listAllSubsidiaryLocales(): Promise<SubsidiaryLocale[]> {
  return AppDataSource.getRepository(SubsidiaryLocale).find({
    order: { subsidiaryName: "ASC", sortOrder: "ASC" },
  });
}

/** One subsidiary's own allowed locale list, fallback first — the shape both
 * admin's Form Initiator convenience picker and a subsidiary user's own ad-hoc
 * locale picker consume. */
export function listSubsidiaryLocales(subsidiaryName: string): Promise<SubsidiaryLocale[]> {
  return AppDataSource.getRepository(SubsidiaryLocale).find({
    where: { subsidiaryName },
    order: { isFallback: "DESC", sortOrder: "ASC" },
  });
}

export interface CreateSubsidiaryLocaleInput {
  subsidiaryName: string;
  code: string;
  langSubtag: string;
  isRtl: boolean;
  label: string;
  isFallback: boolean;
}

/** Adds one locale to a subsidiary's master list. Rejects an exact-duplicate code
 * for that subsidiary with a 409, same convention as
 * subsidiaryProjectBlockService.createSubsidiaryProjectBlock. When `isFallback` is
 * true, first clears any other fallback for this subsidiary — exactly one row per
 * subsidiary is the designated fallback (enforced here, not the DB), mirroring
 * LocaleManagerPanel's own "first locale = fallback" convention. */
export async function addSubsidiaryLocale(input: CreateSubsidiaryLocaleInput): Promise<SubsidiaryLocale> {
  const repo = AppDataSource.getRepository(SubsidiaryLocale);
  const existing = await repo.findOne({ where: { subsidiaryName: input.subsidiaryName, code: input.code } });
  if (existing) {
    throw new ConflictError(`"${input.code}" is already on ${input.subsidiaryName}'s locale list`);
  }

  if (input.isFallback) {
    await repo.update({ subsidiaryName: input.subsidiaryName }, { isFallback: false });
  }

  const count = await repo.count({ where: { subsidiaryName: input.subsidiaryName } });
  return repo.save(repo.create({ ...input, sortOrder: count }));
}

/** Removes one locale from a subsidiary's master list. Returns false if it didn't
 * exist, callers map that to a 404 — same convention as
 * subsidiaryProjectBlockService.deleteSubsidiaryProjectBlock. */
export async function removeSubsidiaryLocale(id: string): Promise<boolean> {
  const result = await AppDataSource.getRepository(SubsidiaryLocale).delete({ id });
  return (result.affected ?? 0) > 0;
}

File: backend-py/app/services/subsidiary_locale_service.py
"""Port of `backend/src/services/subsidiaryLocaleService.ts`.

Removing a locale is a soft delete (`SubsidiaryLocale.isDeleted`): the row is
kept and hidden, and adding the same code to the same subsidiary again restores
it (the `(subsidiaryName, code)` pair is unique in the DB).
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone

from sqlalchemy import func, select, update
from sqlalchemy.orm import Session

from app.errors import ConflictError
from app.models.subsidiary_locale import SubsidiaryLocale


def list_all_subsidiary_locales(db: Session) -> list[SubsidiaryLocale]:
    """Every master locale row, every subsidiary — the admin management
    view."""
    return list(
        db.execute(
            select(SubsidiaryLocale)
            .where(SubsidiaryLocale.isDeleted == False)  # noqa: E712
            .order_by(SubsidiaryLocale.subsidiaryName.asc(), SubsidiaryLocale.sortOrder.asc())
        ).scalars().all()
    )


def list_subsidiary_locales(db: Session, subsidiary_name: str) -> list[SubsidiaryLocale]:
    """One subsidiary's own allowed locale list, fallback first."""
    return list(
        db.execute(
            select(SubsidiaryLocale)
            .where(SubsidiaryLocale.subsidiaryName == subsidiary_name, SubsidiaryLocale.isDeleted == False)  # noqa: E712
            .order_by(SubsidiaryLocale.isFallback.desc(), SubsidiaryLocale.sortOrder.asc())
        ).scalars().all()
    )


@dataclass
class CreateSubsidiaryLocaleInput:
    subsidiaryName: str
    code: str
    langSubtag: str
    isRtl: bool
    label: str
    isFallback: bool


def add_subsidiary_locale(db: Session, input: CreateSubsidiaryLocaleInput) -> SubsidiaryLocale:
    """Adds one locale to a subsidiary's master list. Rejects an
    exact-duplicate code for that subsidiary with a `ConflictError`. When
    `isFallback` is true, first clears any other fallback for this
    subsidiary — exactly one row per subsidiary is the designated
    fallback (enforced here, not the DB). A code that was previously removed
    (soft-deleted) is restored with the new details instead of inserted again."""
    existing = db.execute(
        select(SubsidiaryLocale).where(
            SubsidiaryLocale.subsidiaryName == input.subsidiaryName, SubsidiaryLocale.code == input.code
        )
    ).scalar_one_or_none()
    if existing is not None and not existing.isDeleted:
        raise ConflictError(f'"{input.code}" is already on {input.subsidiaryName}\'s locale list')

    if input.isFallback:
        db.execute(
            update(SubsidiaryLocale)
            .where(SubsidiaryLocale.subsidiaryName == input.subsidiaryName)
            .values(isFallback=False)
        )

    count = db.execute(
        select(func.count()).select_from(SubsidiaryLocale).where(
            SubsidiaryLocale.subsidiaryName == input.subsidiaryName, SubsidiaryLocale.isDeleted == False  # noqa: E712
        )
    ).scalar_one()

    if existing is not None:
        existing.langSubtag = input.langSubtag
        existing.isRtl = input.isRtl
        existing.label = input.label
        existing.isFallback = input.isFallback
        existing.sortOrder = count
        existing.isDeleted = False
        existing.deletedAt = None
        created = existing
    else:
        created = SubsidiaryLocale(
            subsidiaryName=input.subsidiaryName,
            code=input.code,
            langSubtag=input.langSubtag,
            isRtl=input.isRtl,
            label=input.label,
            isFallback=input.isFallback,
            sortOrder=count,
        )
    db.add(created)
    db.commit()
    db.refresh(created)
    return created


def remove_subsidiary_locale(db: Session, id: str) -> bool:
    """Soft-removes one locale from a subsidiary's master list. Returns `False`
    if it didn't exist (or was already removed) — callers map that to a 404."""
    existing = db.get(SubsidiaryLocale, id)
    if existing is None or existing.isDeleted:
        return False
    existing.isDeleted = True
    existing.deletedAt = datetime.now(timezone.utc)
    existing.isFallback = False
    db.add(existing)
    db.commit()
    return True
